### PR TITLE
quantum-espresso: 6.6 -> 7.2

### PR DESCRIPTION
**Pull request by sheepforce**
## Description of changes

Quantum-espresso update, skipping many versions in between: https://www.quantum-espresso.org/release-notes/release-notes-QE7-2.html

Also adds [wannier90](https://github.com/wannier-developers/wannier90) and [libmbd](https://github.com/libmbd/libmbd) as dependencies of the new quantum-espresso versions. Some other dependencies are built via Git submodules by quantum-espresso's cmake on the fly.

I wasn't able to add the GPU support as it requires the NVidia Fortran compiler from the Cuda HPC SDK, which is not yet in nixpkgs.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by markuskowa — CHANGES_REQUESTED** [review]
* quantum espresso has a hash mismatch
* wannier90 fails some tests on aarch64-linux

File: pkgs/by-name/li/libmbd/package.nix [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    cat > cmake/libMBDVersionTag.cmake << EOF
+      set(VERSION_TAG "${version}")
+    EOF
+  '';
+
+  nativeBuildInputs = [ cmake gfortran ];
+
+  buildInputs = [ blas lapack scalapack ];
+
+  propagatedBuildInputs = [mpi];
```
Comment (markuskowa) [inline pkgs/by-name/li/libmbd/package.nix:36]:
```suggestion
  propagatedBuildInputs = [ mpi ];
```

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -38866,9 +38866,12 @@ with pkgs;
 
   pymol = callPackage ../applications/science/chemistry/pymol { };
 
-  quantum-espresso = callPackage ../applications/science/chemistry/quantum-espresso { };
-
-  quantum-espresso-mpi = callPackage ../applications/science/chemistry/quantum-espresso { useMpi = true; };
+  quantum-espresso = callPackage ../applications/science/chemistry/quantum-espresso {
+    hdf5 = hdf5.override {
```
Comment (markuskowa) [inline pkgs/top-level/all-packages.nix:38870]: This is already defined ad `hdf5-fortran`. You could just use `hdf5-fortran` directly as an input for quantum-espresso

File: pkgs/by-name/li/libmbd/package.nix [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+  meta = with lib; {
+    description = " Many-body dispersion library";
+    homepage = "https://github.com/libmbd/libmbd";
+    license = licenses.mpl20;
+    # platforms = platforms.linux;
```
Comment (markuskowa) [inline pkgs/by-name/li/libmbd/package.nix:42]:
```suggestion
    platforms = platforms.linux;
```
?

**Comment by sheepforce** [comment]
Fixed the Quantum-Espresso hash and restricted Wannier90 to x86-64-linux.

File: pkgs/applications/science/chemistry/quantum-espresso/default.nix [hunk 4]
```diff
@@ -1,41 +1,80 @@
...
-  src = fetchFromGitLab {
-    owner = "QEF";
-    repo = "q-e";
-    rev = "qe-${version}";
-    sha256 = "1mkfmw0fq1dabplzdn6v1abhw0ds55gzlvbx3a9brv493whk21yp";
-  };
-
-  passthru = {
-    inherit mpi;
+  # Fetch submodule dependencies. The git tag for the release can not be used
```
Comment (markuskowa) [inline pkgs/applications/science/chemistry/quantum-espresso/default.nix:28]: I think you leave it at `fetchFromGitLab`. The moment you add `fetchSubmodules`, `fetchgit` is automatically used as a backend: https://github.com/NixOS/nixpkgs/blob/af69bf31f1952ce944dfacf541d3b4bef9faed78/pkgs/build-support/fetchgitlab/default.nix#L16

Comment (sheepforce) [reply pkgs/applications/science/chemistry/quantum-espresso/default.nix:28]: Hm but then `fetchFromGitLab` also doesn't recognise the tag and requires the additional argument `leaveDotGit = true`.

Comment (sheepforce) [reply pkgs/applications/science/chemistry/quantum-espresso/default.nix:28]: Uff, that is bad :/ I'm trying right now with the release tarball. Unfortunately all Git submodules are missing, which was the original reason to use the fetchgit fetcher ... I guess I will try to place them manually then ... :(

Comment (sheepforce) [reply pkgs/applications/science/chemistry/quantum-espresso/default.nix]: I'm manually initialising submodules now. Should hopefully work now.

File: pkgs/applications/science/chemistry/quantum-espresso/default.nix [hunk 5]
```diff
@@ -1,41 +1,118 @@
...
 , mpi
 }:
 
+assert ! blas.isILP64;
+assert ! lapack.isILP64;
+
+let
+  # "rev"s must exactly match the git submodule commits in the QE repo
+  gitSubmodules = {
+    devxlib = fetchgit {
```
[the remark this answers is not in the document]
Comment (markuskowa) [reply pkgs/applications/science/chemistry/quantum-espresso/default.nix:30]: It actually should support `group`:
https://github.com/NixOS/nixpkgs/blob/af69bf31f1952ce944dfacf541d3b4bef9faed78/pkgs/build-support/fetchgitlab/default.nix#L5

[the remark this answers is not in the document]
Comment (sheepforce) [reply pkgs/applications/science/chemistry/quantum-espresso/default.nix]: Hu, it does! :smile: Buuuut somehow wrong. In this case I have to give the owner as the group and the group as the owner or the URL is wrong :sweat_smile:  Anyway; changed.

File: pkgs/by-name/li/libmbd/package.nix [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  '';
+
+  nativeBuildInputs = [ cmake gfortran ];
+
+  buildInputs = [ blas lapack scalapack ];
+
+  propagatedBuildInputs = [ mpi ];
+
+  meta = with lib; {
+    description = " Many-body dispersion library";
```
Comment (markuskowa) [inline pkgs/by-name/li/libmbd/package.nix:39]:
```suggestion
    description = "Many-body dispersion library";
```

**Review by markuskowa — APPROVED** [review]

**Pull request merged**